### PR TITLE
Update TSM compaction logic

### PR DESCRIPTION
**Pull request by pauldix**
This fixes up all the compaction logic. However, if you're running a full stress it usually gets too bogged down to compact past the first step size. But if you kill the stress it grabs the next files and compacts them immediately. Next I'll look into compacting multiple levels at the same time to remove that bottleneck.

* Update compaction to look at newest files of the smallest step first
* Update compaction to look at older files in larger steps if newer files don't have enough small steps to compact
* Changed the TestDefaultCompactionPlanner_CombineSequence test to reflect what's possible now. We'd only have multiple files in the same generation if the all files but one were over the max allowable size.
* Clean up the logic on when full compactions are run and when planning can be skipped

cc @jwilder @otoolep

### CODE REVIEW

File: tsdb/engine/tsm1/compact.go [hunk 1]
```diff
@@ -156,47 +134,71 @@ func (c *DefaultPlanner) Plan(lastWrite time.Time) []string {
...
+		return tsmFiles
 	}
 
-	// The set of files that that should be compacted
-	compacted := &tsmGeneration{}
-	var genCount int
-	for _, gen := range order {
-		group := generations[gen]
+	// don't plan if nothing has changed in the filestore
+	if c.lastPlanCheck.UnixNano() > c.FileStore.LastModified().UnixNano() {
```
Comment (jwilder) [inline tsdb/engine/tsm1/compact.go]: `c.lastPlanCheck.After(c.FileStore.LastModified())`

File: tsdb/engine/tsm1/compact.go [hunk 2]
```diff
@@ -205,12 +207,14 @@ func (c *DefaultPlanner) Plan(lastWrite time.Time) []string {
 		return nil
 	}
 
+	c.lastPlanCompactedFull = false
+
 	return tsmFiles
```
Comment (jwilder) [inline tsdb/engine/tsm1/compact.go]: To compact multiple sets concurrently, I was thinking this return value could be changed return a slice of `CompactionGroup` next.  That way `Plan` is still run serially, but each `CompactionGroup` could be run concurrently if needed.

**Comment by jwilder** [comment]
One small thing, but looks good.  :+1:

File: tsdb/engine/tsm1/compact.go (hunk 1 above)
Comment (pauldix) [inline tsdb/engine/tsm1/compact.go]: hah, much better :)

File: tsdb/engine/tsm1/compact.go (hunk 2 above)
Comment (pauldix) [inline tsdb/engine/tsm1/compact.go]: yeah, exactly my thought. Then we could run multiple levels at the same time. Or if there are many in the same level we could split it in two and run two compactions at the same time.

**Comment by pauldix** [comment]
cherry-picked to 0.9.6

**Pull request merged**